File: api/src/functions/carbon-footprint.js
const { app, output, input } = require('@azure/functions');

const tableOutput = output.table({
	tableName: 'CarbonFootprint',
	connection: 'INTERACTIONS_STORAGE_CONNECTION_STRING',
});

app.http('carbon-footprint', {
	methods: ['POST'],
	authLevel: 'function',
	extraOutputs: [tableOutput],
	handler: async (request, context) => {
		context.log(`Http function processed request for url "${request.url}"`);

		try {
			const requestBody = await request.text();
			const requestParams = new URLSearchParams(requestBody);

			const userInputLength = requestParams.get("ul");
			const userInputWordCount = requestParams.get("uw");
			const simplifiedOutputLength = requestParams.get("sl");
			const simplifiedOutputWordCount = requestParams.get("sw");
			const duration = requestParams.get("d");
			const interactionId = requestParams.get("i");

			if ([userInputLength, userInputWordCount, simplifiedOutputLength, simplifiedOutputWordCount, duration, interactionId].some(param => !param)) {
				return {
					status: 400,
					headers: { "Content-Type": "text/plain;charset=utf-8" },
					body: "Missing parameters.",
				};
			}

			const rows = [];
			rows.push({
				PartitionKey: 'Carbon',
				RowKey: context.invocationId,
				InteractionId: interactionId,
				UserInputLength: userInputLength,
				UserInputWordCount: userInputWordCount,
				SimplifiedOutputLength: simplifiedOutputLength,
				SimplifiedOutputWordCount: simplifiedOutputWordCount,
				Duration: duration,
			});
			context.extraOutputs.set(tableOutput, rows);

			return { status: 201 };
		} catch (error) {
			context.error(error);

			return {
				status: 503,
				headers: { "Content-Type": "text/plain;charset=utf-8" },
				body: "Service has failed to process the request. Please try again later.",
			};
		}
	}
});

File: api-dotnet/Functions/CarbonFootprint.cs
using Azure.Data.Tables;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.Extensions.Logging;
using System.Net;

namespace EasyToRead.Api.Functions;

public class CarbonFootprint
{
    private readonly ILogger<CarbonFootprint> _logger;

    public CarbonFootprint(ILogger<CarbonFootprint> logger)
    {
        _logger = logger;
    }

    [Function("carbon-footprint")]
    public async Task<HttpResponseData> Run(
        [HttpTrigger(AuthorizationLevel.Function, "post", Route = "carbon-footprint")] HttpRequestData req,
        FunctionContext context)
    {
        _logger.LogInformation("Http function processed request for url \"{Url}\"", req.Url);

        try
        {
            var requestBody = await req.ReadAsStringAsync() ?? string.Empty;
            var requestParams = FormDataParser.Parse(requestBody);

            var userInputLength = requestParams.GetValueOrDefault("ul");
            var userInputWordCount = requestParams.GetValueOrDefault("uw");
            var simplifiedOutputLength = requestParams.GetValueOrDefault("sl");
            var simplifiedOutputWordCount = requestParams.GetValueOrDefault("sw");
            var duration = requestParams.GetValueOrDefault("d");
            var interactionId = requestParams.GetValueOrDefault("i");

            if (new[] { userInputLength, userInputWordCount, simplifiedOutputLength, simplifiedOutputWordCount, duration, interactionId }
                .Any(string.IsNullOrEmpty))
            {
                var badRequest = req.CreateResponse(HttpStatusCode.BadRequest);
                badRequest.Headers.Add("Content-Type", "text/plain;charset=utf-8");
                await badRequest.WriteStringAsync("Missing parameters.");
                return badRequest;
            }

            var connectionString = Environment.GetEnvironmentVariable("INTERACTIONS_STORAGE_CONNECTION_STRING") ?? string.Empty;
            var tableClient = new TableClient(connectionString, "CarbonFootprint");
            await tableClient.CreateIfNotExistsAsync();

            var entity = new TableEntity("Carbon", context.InvocationId)
            {
                ["InteractionId"] = interactionId,
                ["UserInputLength"] = userInputLength,
                ["UserInputWordCount"] = userInputWordCount,
                ["SimplifiedOutputLength"] = simplifiedOutputLength,
                ["SimplifiedOutputWordCount"] = simplifiedOutputWordCount,
                ["Duration"] = duration,
            };
            await tableClient.AddEntityAsync(entity);

            return req.CreateResponse(HttpStatusCode.Created);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing carbon-footprint request");

            var errorResponse = req.CreateResponse(HttpStatusCode.ServiceUnavailable);
            errorResponse.Headers.Add("Content-Type", "text/plain;charset=utf-8");
            await errorResponse.WriteStringAsync("Service has failed to process the request. Please try again later.");
            return errorResponse;
        }
    }
}
